Add support for some symbolic function expressions

glguy opened the pull request.

glguy committed:
Add support for some symbolic function expressions

glguy committed:
fixup: Simplification and comments

RyanGlScott commented on rme-what4/src/Data/RME/What4.hs:
```diff
@@ -59,29 +64,65 @@ rmeAdapterCheckSat _ logger asserts k =
+-- was actually consistent in a satisfying model. It is possible that a
+-- function can have different outputs for the same concrete inputs.
+--
+-- For example, if we find `f a` and `f b` in the term the system
```
```suggestion
-- For example, if we find `f a` and `f b` in the term, the system
```

RyanGlScott commented on rme-what4/src/Data/RME/What4.hs:
```diff
@@ -59,29 +64,65 @@ rmeAdapterCheckSat _ logger asserts k =
+--
+-- For example, if we find `f a` and `f b` in the term the system
+-- will assign each of those fresh variables not knowing if they
+-- if `a` and `b` might eventually be equal.
```
```suggestion
-- will assign each of those fresh variables not knowing if
-- `a` and `b` might eventually be equal.
```

RyanGlScott commented on rme-what4/src/Data/RME/What4.hs:
```diff
@@ -1,4 +1,5 @@
-{-# Language LambdaCase, GADTs, ImportQualifiedPost, BlockArguments, TypeFamilies, RankNTypes #-}
+{-# Language LambdaCase, GADTs, ImportQualifiedPost, BlockArguments, TypeFamilies, RankNTypes, PatternSynonyms, TypeOperators, MonadComprehensions #-}
+{-# LANGUAGE ScopedTypeVariables #-}
```
Minor style nit pick: it would be nice to use consistent `LANGUAGE` pragma conventions throughout—either having a single line with all of the language pragmas, or putting each pragma on its own line. (I would favor the latter convention, as it tends to be more common in practice.)

glguy replied:
This straggler is a result of editor automation. I'll fix it. I think that most of these language pragmas are noise and don't deserve to fill the screen when the file is first opened. I'll move them out to the cabal file.

RyanGlScott commented on rme-what4/src/Data/RME/What4.hs:
```diff
@@ -20,6 +21,10 @@ import Control.Monad (replicateM, ap, (<$!>))
 import Data.BitVector.Sized qualified as BV
 import Data.IntSet (IntSet)
 import Data.IntSet qualified as IntSet
+import Data.Parameterized (TestEquality(..), TraversableFC(..), (::>), OrdF (compareF), OrderingF (..), type (:~:)(Refl), lexCompareF, joinOrderingF, fromOrdering, Pair (Pair))
```
Consider adding some newlines here to break up this overly long line.

RyanGlScott commented on rme-what4/src/Data/RME/What4.hs:
```diff
@@ -59,29 +64,65 @@ rmeAdapterCheckSat _ logger asserts k =
           putStrLn e
           k W4.Unknown
-      Right (rme, s) ->
+      Right (rme, s :: S t) ->
```
It might be worth binding all of the type variables in the type signature with a top-level `forall` so that we don't have to bring `t` into scope in a random spot in the body of the function using a pattern signature.

RyanGlScott commented on rme-what4/src/Data/RME/What4.hs:
```diff
@@ -59,29 +64,65 @@ rmeAdapterCheckSat _ logger asserts k =
+    same BitRepr (W4.GVW x) (W4.GVW y) = x == y
+    same BVRepr{} (W4.GVW x) (W4.GVW y) = x == y
+
+-- | Given a satisfying model compute the ground value of an RME term.
```
```suggestion
-- | Given a satisfying model, compute the ground value of an RME term.
```

RyanGlScott commented on rme-what4/CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 
+## 0.1.1
+
+* Added support for symbolic functions
```
Does "symbolic function" mean the same thing as "uninterpreted function" here? I am more used to seeing the latter phrase than the former, so I wonder if we should use that here as well.

RyanGlScott replied:
On second thought, What4 also uses the terminology "symbolic function", so I retract my suggestion here.

RyanGlScott commented on rme-what4/src/Data/RME/What4.hs:
```diff
@@ -111,14 +152,81 @@ set s = M (\_ _ t -> t () s)
+
+instance TestEquality (AbstractKey t) where
+  testEquality (AbstractKey f ts_ _ xs_) (AbstractKey g _ _ ys_) =
+    [Refl | Refl <- testEquality f g, Refl <- go ts_ xs_ ys_]
```
It took me a while to figure out that this is defined using a monad comprehension where `Maybe` is the `Monad`. It might be worth leaving a brief comment saying as much. (Similarly for the `TestEquality (ConcreteKey t)` instance.)

RyanGlScott commented on rme-what4/src/Data/RME/What4.hs:
```diff
@@ -137,14 +245,14 @@ type family R (t :: W4.BaseType) where
 
 -- | Newtype wrapper for 'R' type for use with 'MapF'
-newtype SomeR tp = SomeR (R tp)
+newtype R' tp = R (R tp)
```
Any reason not to use the name `R'` for both the newtype name and the newtype constructor? Using both `R'` and `R` here is a bit confusing.

glguy replied:
With one constructor it's a bit arbitrary if we have the constructor describe the type wrapped or the newtype wrapper. In this case there's no other need for the data constructor `R` so it felt like using `R'` would just add meaningless noise to the various pattern matches around the file.

RyanGlScott replied:
Fair enough. I suppose this matches the `RV'`/`RV` convention used in `crucible` as well. (I don't really like this convention, but I don't have a suggestion for something better.)

glguy committed:
Improvements due to code review

glguy committed:
Call the function uninterpreted

glguy committed:
Use CEGAR to resolve uninterpreted function contradictions

Instead of aborting when the model selections a contradictory interpretation of an uninterpreted function symbol, this change compute the symbolic deconfliction on demand and rechecks satisfiability.

glguy committed:
Remove debugging code

glguy committed:
Cleanup and documentation
